source/slang/slang-ast-serialize.cpp

```diff
@@ -0,0 +1,1289 @@
+
+// String
+template <>
+struct ASTSerialTypeInfo<String>
+{
+    typedef String NativeType;
+    typedef ASTSerialIndex SerialType;
+    enum { SerialAlignment = SLANG_ALIGN_OF(SerialType) };
+
+    static void toSerial(ASTSerialWriter* writer, const void* native, void* serial)
+    {
+        auto& src = *(const NativeType*)native;
+        *(SerialType*)serial = writer->addString(src);
+    }
+    static void toNative(ASTSerialReader* reader, const void* serial, void* native)
+    {
+        auto& src = *(const SerialType*)serial;
+        auto& dst = *(NativeType*)native;
+        dst = reader->getString(src);
+    }
+};
+
+// Token
+template <>
+struct ASTSerialTypeInfo<Token>
+{
+    typedef Token NativeType;
+    struct SerialType
+    {
+        ASTSerialTypeInfo<BaseType>::SerialType type;
+        ASTSerialTypeInfo<SourceLoc>::SerialType loc;
+        ASTSerialIndex name;
+    };
+    enum { SerialAlignment = SLANG_ALIGN_OF(SerialType) };
+
+    static void toSerial(ASTSerialWriter* writer, const void* native, void* serial)
+    {
+        auto& src = *(const NativeType*)native;
+        auto& dst = *(SerialType*)serial;
+
+        ASTSerialTypeInfo<TokenType>::toSerial(writer, &src.type, &dst.type);
+        ASTSerialTypeInfo<SourceLoc>::toSerial(writer, &src.loc, &dst.loc);
+        dst.name = writer->addName(src.getName());
+    }
+    static void toNative(ASTSerialReader* reader, const void* serial, void* native)
+    {
+        auto& src = *(const SerialType*)serial;
+        auto& dst = *(NativeType*)native;
+
+        dst.flags = 0;
+        dst.charsNameUnion.chars = nullptr;
+
+        ASTSerialTypeInfo<TokenType>::toNative(reader, &src.type, &dst.type);
+        ASTSerialTypeInfo<SourceLoc>::toNative(reader, &src.loc, &dst.loc);
+
+        if (src.name != ASTSerialIndex(0))
+        {
+            dst.charsNameUnion.name = reader->getName(src.name);
+            dst.flags |= TokenFlag::Name;
+        }
+    }
+};
+
+// NameLoc
+template <>
+struct ASTSerialTypeInfo<NameLoc>
+{
+    typedef NameLoc NativeType;
+    struct SerialType
+    {
+        ASTSerialTypeInfo<SourceLoc>::SerialType loc;
+        ASTSerialIndex name;
+    };
+    enum { SerialAlignment = SLANG_ALIGN_OF(SerialType) };
+
+    static void toSerial(ASTSerialWriter* writer, const void* native, void* serial)
+    {
+        auto& src = *(const NativeType*)native;
+        auto& dst = *(SerialType*)serial;
+
+        dst.name = writer->addName(src.name);
+        ASTSerialTypeInfo<SourceLoc>::toSerial(writer, &src.loc, &dst.loc);
+    }
+    static void toNative(ASTSerialReader* reader, const void* serial, void* native)
+    {
+        auto& src = *(const SerialType*)serial;
+        auto& dst = *(NativeType*)native;
+
+        dst.name = reader->getName(src.name);
+        ASTSerialTypeInfo<SourceLoc>::toNative(reader, &src.loc, &dst.loc);
+    }
+};
+
+// DiagnosticInfo
+template <>
+struct ASTSerialTypeInfo<const DiagnosticInfo*>
```

Comment:
Oh yeah, this is required because we have a modifier that can be used to suppress a specific diagnostic by name. I think that serializing these by name is the right call here, since it should be (in theory) more future-proof.